migrations/20210612_initial.sql

```diff
@@ -0,0 +1,104 @@
+CREATE TABLE profiles (
+  profile_id INTEGER NOT NULL PRIMARY KEY,
+  display_name TEXT NOT NULL,
+  properties TEXT NOT NULL -- JSON with user or contact profile
+);
+
+CREATE TABLE known_servers(
+  host TEXT NOT NULL,
+  port TEXT NOT NULL,
+  key_hash BLOB,
+  PRIMARY KEY (host, port)
+) WITHOUT ROWID;
+
+CREATE TABLE contacts (
+  contact_id INTEGER NOT NULL PRIMARY KEY,
+  display_name TEXT NOT NULL, -- display name set by local user
+  properties TEXT NOT NULL, -- JSON set by local user
+  profile_id INTEGER REFERENCES profiles, -- sent by remote contact
+  contact_status TEXT NOT NULL DEFAULT ''
+);
+
+CREATE TABLE contact_connections (
+  connection_id BLOB NOT NULL PRIMARY KEY,
+  contact_id INTEGER NOT NULL REFERENCES contacts ON DELETE RESTRICT,
+  conn_level INTEGER NOT NULL DEFAULT 0,
+  via_conn BLOB REFERENCES contact_connections (connection_id),
+  conn_status TEXT NOT NULL DEFAULT ''
+) WITHOUT ROWID;
+
+CREATE TABLE contact_invitations (
+  invitation_id BLOB NOT NULL PRIMARY KEY,
+  contact_id INTEGER NOT NULL REFERENCES contacts ON DELETE RESTRICT,
+  invitation_status TEXT NOT NULL DEFAULT ''
+) WITHOUT ROWID;
+
+CREATE TABLE groups (
+  group_id INTEGER NOT NULL PRIMARY KEY, -- local group ID
+  display_name TEXT NOT NULL, -- shared
+  properties TEXT NOT NULL -- shared JSON group profile
+);
+
+CREATE TABLE group_members ( -- group members, nullable fields are NULL for the local user
+  group_id INTEGER NOT NULL REFERENCES groups ON DELETE RESTRICT,
+  member_id BLOB NOT NULL, -- shared member ID, unique per group
+  member_role TEXT NOT NULL DEFAULT '', -- owner, admin, moderator, ''
+  member_status TEXT NOT NULL DEFAULT '', -- inv | con | full | off
+  contact_id INTEGER REFERENCES contacts ON DELETE RESTRICT,
```

Comment:
`NOT NULL`?

Reply:
that's actually on purpose to include the current user membership in the same table... Maybe current user should simply be present in contacts table with some boolean flag (or just ID = 0 or 1).

Initially I repeated all member properties for the current user in groups table, but it was messy - there are many fields

Reply:
I wouldn't mix user's profile with contacts' profiles at all - why not have a separate table "`user_data`" for that or like you offered a configuration file. The downside is it'd be the only record in the table. Then again user can have more configuration in his chat and his settings have to be stored somewhere